wlb: T4470: Migrate WAN Load Balancer to XML/Python

sarthurdev opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
This PR migrates the last remaining Vyatta perl module to XML/Python standard.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [x] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
wanloadbalance

## Proposed changes
<!--- Describe your changes in detail -->
* Migrate perl scripts/daemon to Python
* Migrate op-mode from vyatta-wanloadbalance package

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_load-balancing_wan.py
DEBUG - test_check_chains (__main__.TestLoadBalancingWan.test_check_chains) ... ok
DEBUG - test_table_routes (__main__.TestLoadBalancingWan.test_table_routes) ... ok
DEBUG - 
DEBUG - ----------------------------------------------------------------------
DEBUG - Ran 2 tests in 16.750s
DEBUG - 
DEBUG - OK
```


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

sarthurdev committed:
wlb: T4470: Migrate WAN load balancer to Python/XML

sever-sever commented on python/vyos/wanloadbalance.py:
```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2022 VyOS maintainers and contributors
```
Should be 2024

sever-sever commented on src/helpers/vyos-load-balancer.py:
```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/python3
+
+# Copyright 2022 VyOS maintainers and contributors <maintainers@vyos.io>
```
2024

sarthurdev committed:
wlb: T4470: Support WLB op-mode commands

sever-sever commented on python/vyos/wanloadbalance.py:
```diff
@@ -0,0 +1,133 @@
+    rc = run(cmd_str)
+    return rc != 0
+
+def parse_dhcp_nexthop(ifname):
```
Will PPPoE interfaces also work?
I didn't find any code regarding PPP.

sarthurdev replied:
Good question, let me look into that.

sarthurdev replied:
i've looked how the old vyatta module handled it, it shouldn't be too difficult. I'll work on it after I've got all reported bugs sorted.

sever-sever commented on src/helpers/vyos-load-balancer.py:
```diff
@@ -0,0 +1,262 @@
+                run(f'ip route replace table {table_num} default dev {ifname} via {health_conf["nexthop"]}')
+
+            run(f'ip route delete table {table_num}')
+            run(f'ip rule add fwmark {hex(table_num)} table {table_num}')
```
The `ip rule`  are not deleted
```
vyos@router:~$ show conf com | match wan
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
vyos@router:~$ 
vyos@router:~$ 
vyos@router:~$ ip rule
0:	from all lookup local
32756:	from all fwmark 0xca lookup 202
32757:	from all fwmark 0xc9 lookup 201
32758:	from all fwmark 0xca lookup 202
32759:	from all fwmark 0xc9 lookup 201
32760:	from all fwmark 0xca lookup 202
32761:	from all fwmark 0xc9 lookup 201
32762:	from all fwmark 0xca lookup 202
32763:	from all fwmark 0xc9 lookup 201
32764:	from all fwmark 0xca lookup 202
32765:	from all fwmark 0xc9 lookup 201
32766:	from all lookup main
32767:	from all lookup default
vyos@router:~$ 

vyos@router# delete load-balancing 
[edit]
vyos@router# commit
[edit]
vyos@router# ip rule
0:	from all lookup local
32756:	from all fwmark 0xca lookup 202
32757:	from all fwmark 0xc9 lookup 201
32758:	from all fwmark 0xca lookup 202
32759:	from all fwmark 0xc9 lookup 201
32760:	from all fwmark 0xca lookup 202
32761:	from all fwmark 0xc9 lookup 201
32762:	from all fwmark 0xca lookup 202
32763:	from all fwmark 0xc9 lookup 201
32764:	from all fwmark 0xca lookup 202
32765:	from all fwmark 0xc9 lookup 201
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@router# 

vyos@router# set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
[edit]
vyos@router# set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
[edit]
vyos@router# commit
[edit]
vyos@router# ip rule
0:	from all lookup local
32746:	from all fwmark 0xca lookup 202
32747:	from all fwmark 0xc9 lookup 201
32748:	from all fwmark 0xca lookup 202
32749:	from all fwmark 0xc9 lookup 201
32750:	from all fwmark 0xca lookup 202
32751:	from all fwmark 0xc9 lookup 201
32752:	from all fwmark 0xca lookup 202
32753:	from all fwmark 0xc9 lookup 201
32754:	from all fwmark 0xca lookup 202
32755:	from all fwmark 0xc9 lookup 201
32756:	from all fwmark 0xca lookup 202
32757:	from all fwmark 0xc9 lookup 201
32758:	from all fwmark 0xca lookup 202
32759:	from all fwmark 0xc9 lookup 201
32760:	from all fwmark 0xca lookup 202
32761:	from all fwmark 0xc9 lookup 201
32762:	from all fwmark 0xca lookup 202
32763:	from all fwmark 0xc9 lookup 201
32764:	from all fwmark 0xca lookup 202
32765:	from all fwmark 0xc9 lookup 201
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@router# 

```

sever-sever commented:
1. ip rules not deleted (after deleting wlb config)
2. Delete stick-conections fails
```
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
set load-balancing wan sticky-connections
commit


vyos@router# delete load-balancing wan sticky-connections 
[edit]
vyos@router# commit
[ load-balancing wan ]
Traceback (most recent call last):
  File "/usr/libexec/vyos/services/vyos-configd", line 136, in run_script
    script.apply(c)
  File "/usr/libexec/vyos//conf_mode/load-balancing_wan.py", line 104, in apply
    cmd(f'systemctl restart {service}')
  File "/usr/lib/python3/dist-packages/vyos/utils/process.py", line 155, in cmd
    raise OSError(code, feedback)
PermissionError: [Errno 1] failed to run command: systemctl restart vyos-wan-load-balance.service
returned: 
exit code: 1

[[load-balancing wan]] failed
Commit failed
[edit]
vyos@router# 

```
3. Reverse protocol match does not work:
```
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
set load-balancing wan rule 20 inbound-interface 'eth2'
set load-balancing wan rule 20 interface eth1
set load-balancing wan rule 20 protocol 'tcp'
set load-balancing wan sticky-connections inbound
commit
set load-balancing wan rule 20 protocol !tcp 
commit
```
rules:
```
        chain wlb_mangle_prerouting {
                type filter hook prerouting priority mangle; policy accept;
                iifname "eth0" ct state new ct mark set 0x000000c9
                iifname "eth1" ct state new ct mark set 0x000000ca
                iifname "eth2" meta l4proto tcp ct state new limit rate 5/second burst 5 packets counter packets 0 bytes 0 jump wlb_mangle_isp_eth1
                iifname "eth2" meta l4proto tcp counter packets 0 bytes 0 meta mark set ct mark
        }

```
4. Delete load-balancing does not flush nft rules
```
vyos@router:~$ show conf com | match load
set load-balancing wan enable-local-traffic
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
set load-balancing wan rule 20 inbound-interface 'eth2'
set load-balancing wan rule 20 interface eth1
set load-balancing wan rule 20 protocol '47'
set load-balancing wan sticky-connections inbound
vyos@router:~$ conf
d[edit]
vyos@router# delete load-balancing 
[edit]
vyos@router# commit
[edit]
vyos@router# 

```
check:
```
vyos@router:~$ sudo nft list table vyos_wanloadbalance
table ip vyos_wanloadbalance {
	chain wlb_nat_postrouting {
		type nat hook postrouting priority srcnat - 1; policy accept;
		ct mark 0x000000c9 counter packets 0 bytes 0 snat to 10.0.1.2
		ct mark 0x000000ca counter packets 1 bytes 328 snat to 10.0.2.2
	}

	chain wlb_mangle_prerouting {
		type filter hook prerouting priority mangle; policy accept;
		iifname "eth0" ct state new ct mark set 0x000000c9
		iifname "eth1" ct state new ct mark set 0x000000ca
		iifname "eth2" meta l4proto gre ct state new limit rate 5/second burst 5 packets counter packets 0 bytes 0 jump wlb_mangle_isp_eth1
		iifname "eth2" meta l4proto gre counter packets 0 bytes 0 meta mark set ct mark
	}

	chain wlb_mangle_output {
		type filter hook output priority mangle; policy accept;
		meta mark != 0x00000000 counter packets 0 bytes 0 accept
		meta l4proto icmp counter packets 6096 bytes 552936 accept
		ip saddr 127.0.0.0/8 ip daddr 127.0.0.0/8 counter packets 0 bytes 0 accept
		oifname != "eth2" meta l4proto gre ct state new limit rate 5/second burst 5 packets counter packets 0 bytes 0 jump wlb_mangle_isp_eth1
		oifname != "eth2" meta l4proto gre counter packets 0 bytes 0 meta mark set ct mark
	}

	chain wlb_mangle_isp_eth0 {
		meta mark set 0x000000c9 ct mark set 0x000000c9 counter packets 0 bytes 0 accept
	}

	chain wlb_mangle_isp_eth1 {
		meta mark set 0x000000ca ct mark set 0x000000ca counter packets 0 bytes 0 accept
	}
}
vyos@router:~$ 
```
5. Delete load-balancing does not flush routes
```
vyos@router# delete load-balancing 
[edit]
vyos@router# commit
[edit]
vyos@router# 
[edit]
vyos@router# 
[edit]
vyos@router# run show ip route table all
Codes: K - kernel route, C - connected, S - static, R - RIP,
       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, F - PBR,
       f - OpenFabric,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

VRF default table 201:
K>* 0.0.0.0/0 [0/0] via 10.0.1.1, eth0, 02:32:50

VRF default table 202:
K>* 0.0.0.0/0 [0/0] via 10.0.2.1, eth1, 02:32:50

```
6. Next-hop DHCP does not add routes in the tables
```
set interfaces ethernet eth0 address 'dhcp'
set interfaces ethernet eth1 address 'dhcp'
set load-balancing wan interface-health eth0 nexthop 'dhcp'
set load-balancing wan interface-health eth1 nexthop 'dhcp'
set load-balancing wan rule 20 inbound-interface 'eth2'
set load-balancing wan rule 20 interface eth1
set load-balancing wan rule 20 protocol 'tcp'
set load-balancing wan sticky-connections inbound

```
Check:
```
vyos@router# run show ip route table all
Codes: K - kernel route, C - connected, S - static, R - RIP,
       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, F - PBR,
       f - OpenFabric,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

VRF default table 254:
S>* 0.0.0.0/0 [210/0] via 10.0.1.1, eth0, weight 1, 00:03:15
  *                   via 10.0.2.1, eth1, weight 1, 00:03:15
C>* 10.0.1.0/24 is directly connected, eth0, 00:03:15
C>* 10.0.2.0/24 is directly connected, eth1, 00:03:15
C>* 100.64.0.0/24 is directly connected, eth2, 00:07:34
C>* 192.168.122.0/24 is directly connected, eth4, 00:07:35
[edit]
vyos@router# 
[edit]
vyos@router# run show ip route table 201
[edit]
vyos@router# run show ip route table 202
[edit]
vyos@router# sudo ip route show table 201
Error: ipv4: FIB table does not exist.
Dump terminated
[edit]
vyos@router# sudo ip route show table 202
Error: ipv4: FIB table does not exist.
Dump terminated
[edit]
vyos@router# 

```

sever-sever commented on src/op_mode/load-balancing_wan.py:
```diff
@@ -0,0 +1,117 @@
+            'ifname': ifname,
+            'status': "active" if if_data['state'] else "failed",
+            'last_change': change_dt.strftime("%Y-%m-%d %H:%M:%S") if change_dt else 'N/A',
+            'last_success': str(now - success_dt) if success_dt else 'N/A',
```
What does the output format mean for `Last Interface Success:`?
```
vyos@router:~$ show wan-load-balance 
Interface: eth0
Status: active
Last Status Change: 2024-11-05 14:21:30
Last Interface Success: 0:03:48.162705
Last Interface Failure: N/A
Interface Failures: 0

Interface: eth1
Status: failed
Last Status Change: 2024-11-05 14:21:35
Last Interface Success: 0:03:53.167816
Last Interface Failure: 0:03:43.160109
Interface Failures: 1

vyos@router:~$ 
```

sarthurdev replied:
HH:MM:SS since last successful health test

HollyGurza commented on src/conf_mode/load-balancing_wan.py:
```diff
@@ -59,86 +52,59 @@ def verify(lb):
-        cmd(f'systemctl restart {systemd_service}')
+    cmd(f'systemctl restart {service}')
 
     call_dependents()
```
I think, call to call_dependents() remains valuable even when the load balancer configuration is empty. It can handle cases where an old configuration has been removed, ensuring all dependencies are correctly updated.

sever-sever commented on python/vyos/wanloadbalance.py:
```diff
@@ -0,0 +1,133 @@
+
+from vyos.utils.process import run
+
+dhclient_lease = '/var/lib/dhcp/dhclient_{0}.lease'
```
This path seems incorrect:
```
vyos@router# ls -la /var/lib/dhcp/
total 8
drwxr-xr-x 2 root root    3 Oct 20  2023 .
drwxr-xr-x 1 root root 4096 Nov  5 13:30 ..
[edit]
vyos@router# 

```
The correct path:
```
vyos@router# cat /run/dhclient/dhclient_eth0.lease
Tue Nov 5 16:19:41 UTC 2024
reason='BOUND'
interface='eth0'
new_expiry='1730909981'
new_dhcp_lease_time='86400'
medium=''
alias_ip_address=''
new_ip_address='10.0.1.10'
new_broadcast_address='10.0.1.255'
new_subnet_mask='255.255.255.0'
new_domain_name=''
new_network_number='10.0.1.0'
new_domain_name_servers='1.1.1.1'
new_routers='10.0.1.1'
new_static_routes=''
new_dhcp_server_identifier='10.0.1.1'
new_dhcp_message_type='5'
old_ip_address=''
old_subnet_mask=''
old_domain_name=''
old_domain_name_servers=''
old_routers=''
old_static_routes=''

```

sarthurdev commented:
Thanks @sever-sever 

>     1. ip rules not deleted (after deleting wlb config) 
Fixed

>     2. Delete stick-conections fails
Fixed

>     3. Reverse protocol match does not work:
Fixed

>     4. Delete load-balancing does not flush nft rules
Fixed

>     5. Delete load-balancing does not flush routes

Fixed

>     6. Next-hop DHCP does not add routes in the tables

Fixed

sever-sever commented:
1. Cannot start with this configuration (eth1 is not WAN). Probably additional check is required
```
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan rule 20 inbound-interface 'eth2'
set load-balancing wan rule 20 interface eth1
set load-balancing wan rule 20 protocol 'all'
set load-balancing wan sticky-connections inbound

```
Log:
```
Dec 25 11:04:20 router python3[4531]:                                                                                                                    ~~~~~~~~~~~~^^^^^^^^
Dec 25 11:04:20 router python3[4531]: KeyError: 'eth1'
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Main process exited, code=exited, status=1/FAILURE
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Failed with result 'exit-code'.
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 4.
Dec 25 11:04:20 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Dec 25 11:04:20 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Dec 25 11:04:20 router python3[4537]: Traceback (most recent call last):
Dec 25 11:04:20 router python3[4537]:   File "/usr/libexec/vyos/vyos-load-balancer.py", line 194, in <module>
Dec 25 11:04:20 router python3[4537]:     nftables_update(lb)
Dec 25 11:04:20 router python3[4537]:   File "/usr/libexec/vyos/vyos-load-balancer.py", line 125, in nftables_update
Dec 25 11:04:20 router python3[4537]:     render(nftables_wlb_conf, 'load-balancing/nftables-wlb.j2', lb)
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/vyos/template.py", line 152, in render
Dec 25 11:04:20 router python3[4537]:     rendered = render_to_string(template, content, formater, location)
Dec 25 11:04:20 router python3[4537]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/vyos/template.py", line 121, in render_to_string
Dec 25 11:04:20 router python3[4537]:     rendered = template.render(content)
Dec 25 11:04:20 router python3[4537]:                ^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1301, in render
Dec 25 11:04:20 router python3[4537]:     self.environment.handle_exception()
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 936, in handle_exception
Dec 25 11:04:20 router python3[4537]:     raise rewrite_traceback_stack(source=source)
Dec 25 11:04:20 router python3[4537]:   File "/usr/share/vyos/templates/load-balancing/nftables-wlb.j2", line 31, in top-level template code
Dec 25 11:04:20 router python3[4537]:     {{ rule_conf | wlb_nft_rule(rule_id, limit=limit, weight=True, health_state=health_state) }}
Dec 25 11:04:20 router python3[4537]: ^^^^^^^^^^^^^^^^^^^^^
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/vyos/template.py", line 785, in wlb_nft_rule
Dec 25 11:04:20 router python3[4537]:     return wlb_nft_rule(rule_conf, rule_id, local, exclude, limit, weight, health_state, action, restore_mark)
Dec 25 11:04:20 router python3[4537]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/vyos/wanloadbalance.py", line 83, in nft_rule
Dec 25 11:04:20 router python3[4537]:     weights, total_weight = wlb_weight_interfaces(rule_conf, health_state)
Dec 25 11:04:20 router python3[4537]:                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/vyos/wanloadbalance.py", line 98, in wlb_weight_interfaces
Dec 25 11:04:20 router python3[4537]:     interfaces = [(ifname, int(if_conf.get('weight', 1))) for ifname, if_conf in rule_conf['interface'].items() if health_state[ifname]['state']]
Dec 25 11:04:20 router python3[4537]:                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 11:04:20 router python3[4537]:   File "/usr/lib/python3/dist-packages/vyos/wanloadbalance.py", line 98, in <listcomp>
Dec 25 11:04:20 router python3[4537]:     interfaces = [(ifname, int(if_conf.get('weight', 1))) for ifname, if_conf in rule_conf['interface'].items() if health_state[ifname]['state']]
Dec 25 11:04:20 router python3[4537]:                                                                                                                    ~~~~~~~~~~~~^^^^^^^^
Dec 25 11:04:20 router python3[4537]: KeyError: 'eth1'
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Main process exited, code=exited, status=1/FAILURE
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Failed with result 'exit-code'.
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 5.
Dec 25 11:04:20 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Start request repeated too quickly.
Dec 25 11:04:20 router systemd[1]: vyos-wan-load-balance.service: Failed with result 'exit-code'.
Dec 25 11:04:20 router systemd[1]: Failed to start vyos-wan-load-balance.service - VyOS WAN Load Balancer.
^C
vyos@router:~$ 

```
If it fails, the rules are not deleted:
```
vyos@router# ip rule
0:	from all lookup local
32761:	from all fwmark 0xc9 lookup 201
32762:	from all fwmark 0xc9 lookup 201
32763:	from all fwmark 0xc9 lookup 201
32764:	from all fwmark 0xc9 lookup 201
32765:	from all fwmark 0xc9 lookup 201
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@router# 

```
Delete load-balancing does not help
```
vyos@router# delete load-balancing 
[edit]
vyos@router# commit
[edit]
vyos@router# ip rule
0:	from all lookup local
32756:	from all fwmark 0xc9 lookup 201
32757:	from all fwmark 0xc9 lookup 201
32758:	from all fwmark 0xc9 lookup 201
32759:	from all fwmark 0xc9 lookup 201
32760:	from all fwmark 0xc9 lookup 201
32761:	from all fwmark 0xc9 lookup 201
32762:	from all fwmark 0xc9 lookup 201
32763:	from all fwmark 0xc9 lookup 201
32764:	from all fwmark 0xc9 lookup 201
32765:	from all fwmark 0xc9 lookup 201
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@router# 

```

2. Something is wrong with `ip rule` logic per adding rules
Before wlb
```
vyos@router# ip rule
0:	from all lookup local
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@router# 
```
Add wlb and two rules (2 commits):
```
set load-balancing wan enable-local-traffic
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
set load-balancing wan rule 20 inbound-interface 'eth2'
set load-balancing wan rule 20 interface eth0
set load-balancing wan rule 20 interface eth1
commit
set load-balancing wan rule 9999 protocol 'all'
commit

```
Check IP rules:
```
vyos@router# ip rule
0:	from all lookup local
32758:	from all fwmark 0xca lookup 202
32759:	from all fwmark 0xc9 lookup 201
32760:	from all fwmark 0xca lookup 202
32761:	from all fwmark 0xc9 lookup 201
32762:	from all fwmark 0xca lookup 202
32763:	from all fwmark 0xc9 lookup 201
32764:	from all fwmark 0xca lookup 202
32765:	from all fwmark 0xc9 lookup 201
32766:	from all lookup main
32767:	from all lookup default
[edit]
vyos@router# 

```
3. Empty `rule 100` breaks the daemon/process:
```
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
set load-balancing wan rule 20 inbound-interface 'eth2'
set load-balancing wan rule 20 interface eth0
set load-balancing wan rule 20 interface eth1
set load-balancing wan rule 100
commit
```
Log
```
Dec 25 12:17:12 router python3[22542]: Traceback (most recent call last):
Dec 25 12:17:12 router python3[22542]:   File "/usr/libexec/vyos/vyos-load-balancer.py", line 194, in <module>
Dec 25 12:17:12 router python3[22542]:     nftables_update(lb)
Dec 25 12:17:12 router python3[22542]:   File "/usr/libexec/vyos/vyos-load-balancer.py", line 125, in nftables_update
Dec 25 12:17:12 router python3[22542]:     render(nftables_wlb_conf, 'load-balancing/nftables-wlb.j2', lb)
Dec 25 12:17:12 router python3[22542]:   File "/usr/lib/python3/dist-packages/vyos/template.py", line 152, in render
Dec 25 12:17:12 router python3[22542]:     rendered = render_to_string(template, content, formater, location)
Dec 25 12:17:12 router python3[22542]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 12:17:12 router python3[22542]:   File "/usr/lib/python3/dist-packages/vyos/template.py", line 121, in render_to_string
Dec 25 12:17:12 router python3[22542]:     rendered = template.render(content)
Dec 25 12:17:12 router python3[22542]:                ^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 12:17:12 router python3[22542]:   File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1301, in render
Dec 25 12:17:12 router python3[22542]:     self.environment.handle_exception()
Dec 25 12:17:12 router python3[22542]:   File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 936, in handle_exception
Dec 25 12:17:12 router python3[22542]:     raise rewrite_traceback_stack(source=source)
Dec 25 12:17:12 router python3[22542]:   File "/usr/share/vyos/templates/load-balancing/nftables-wlb.j2", line 31, in top-level template code
Dec 25 12:17:12 router python3[22542]:     {{ rule_conf | wlb_nft_rule(rule_id, limit=limit, weight=True, health_state=health_state) }}
Dec 25 12:17:12 router python3[22542]: ^^^^^^^^^^^^^^^^^^^^^
Dec 25 12:17:12 router python3[22542]:   File "/usr/lib/python3/dist-packages/vyos/template.py", line 785, in wlb_nft_rule
Dec 25 12:17:12 router python3[22542]:     return wlb_nft_rule(rule_conf, rule_id, local, exclude, limit, weight, health_state, action, restore_mark)
Dec 25 12:17:12 router python3[22542]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 12:17:12 router python3[22542]:   File "/usr/lib/python3/dist-packages/vyos/wanloadbalance.py", line 83, in nft_rule
Dec 25 12:17:12 router python3[22542]:     weights, total_weight = wlb_weight_interfaces(rule_conf, health_state)
Dec 25 12:17:12 router python3[22542]:                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 25 12:17:12 router python3[22542]:   File "/usr/lib/python3/dist-packages/vyos/wanloadbalance.py", line 98, in wlb_weight_interfaces
Dec 25 12:17:12 router python3[22542]:     interfaces = [(ifname, int(if_conf.get('weight', 1))) for ifname, if_conf in rule_conf['interface'].items() if health_state[ifname]['state']]
Dec 25 12:17:12 router python3[22542]:                                                                                  ~~~~~~~~~^^^^^^^^^^^^^
Dec 25 12:17:12 router python3[22542]: KeyError: 'interface'

```

sever-sever reviewed (changes requested):
Something is wrong with adding/deleting `ip rule`
I sent examples to reproduce
Thanks!

sever-sever commented:
Tested on
```
vyos@router:~$ show ver
Version:          VyOS 1.5-rolling-202501140634
Release train:    current
Release flavor:   generic

Built by:         autobuild@vyos.net
Built on:         Tue 14 Jan 2025 06:35 UTC
Build UUID:       17bf5ea7-2b1f-45a9-a631-a500d9fa824b
Build commit ID:  0bd5cb4f5038f5
```

01. Interface in the down state, but the test shows as `active`
```
set interfaces ethernet eth0 address '10.0.1.2/24'
set interfaces ethernet eth0 description 'ISP-01'
set interfaces ethernet eth1 address '10.0.2.2/24'
set interfaces ethernet eth1 description 'ISP-02'

set load-balancing wan hook '/bin/true'
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
```
Actual interface status
```
vyos@router:~$ show int
Codes: S - State, L - Link, u - Up, D - Down, A - Admin Down
Interface    IP Address          MAC                VRF        MTU  S/L    Description
-----------  ------------------  -----------------  -------  -----  -----  -------------
eth0         10.0.1.2/24         0c:67:b0:98:00:00  default   1500  u/D    ISP-01
eth1         10.0.2.2/24         0c:67:b0:98:00:01  default   1500  u/u    ISP-02
```
Unexpected `active` for `eth0`
```
vyos@router:~$ show wan-load-balance 
Interface: eth0
Status: active
Last Status Change: N/A
Last Interface Success: N/A
Last Interface Failure: N/A
Interface Failures: 0

Interface: eth1
Status: active
Last Status Change: N/A
Last Interface Success: N/A
Last Interface Failure: N/A
Interface Failures: 0
```

With such a configuration, the service fails:
```
Jan 17 10:04:35 router systemd[1]: vyos-wan-load-balance.service: Main process exited, code=exited, status=1/FAILURE
Jan 17 10:04:35 router systemd[1]: vyos-wan-load-balance.service: Failed with result 'exit-code'.
Jan 17 10:04:36 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 58.
Jan 17 10:04:36 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 10:04:36 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 10:04:41 router python3[5811]: Traceback (most recent call last):
Jan 17 10:04:41 router python3[5811]:   File "/usr/libexec/vyos/vyos-load-balancer.py", line 265, in <module>
Jan 17 10:04:41 router python3[5811]:     on_state_change(lb, ifname, state['state'])
Jan 17 10:04:41 router python3[5811]:   File "/usr/libexec/vyos/vyos-load-balancer.py", line 81, in on_state_change
Jan 17 10:04:41 router python3[5811]:     run(script_path, env=[f'WLB_INTERFACE_NAME={ifname}', f'WLB_INTERFACE_STATE={state_str}'])
Jan 17 10:04:41 router python3[5811]:   File "/usr/lib/python3/dist-packages/vyos/utils/process.py", line 144, in run
Jan 17 10:04:41 router python3[5811]:     _, code = popen(
Jan 17 10:04:41 router python3[5811]:               ^^^^^^
Jan 17 10:04:41 router python3[5811]:   File "/usr/lib/python3/dist-packages/vyos/utils/process.py", line 104, in popen
Jan 17 10:04:41 router python3[5811]:     p = Popen(command, stdin=stdin, stdout=stdout, stderr=stderr,
Jan 17 10:04:41 router python3[5811]:         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 17 10:04:41 router python3[5811]:   File "/usr/lib/python3.11/subprocess.py", line 1024, in __init__
Jan 17 10:04:41 router python3[5811]:     self._execute_child(args, executable, preexec_fn, close_fds,
Jan 17 10:04:41 router python3[5811]:   File "/usr/lib/python3.11/subprocess.py", line 1817, in _execute_child
Jan 17 10:04:41 router python3[5811]:     for k, v in env.items():
Jan 17 10:04:41 router python3[5811]:                 ^^^^^^^^^
Jan 17 10:04:41 router python3[5811]: AttributeError: 'list' object has no attribute 'items'

```
2. If the service restarts continuously  (as in point 1), it does not clean the rules
```
vyos@router# ip rule | head -8
0:	from all lookup local
32220:	from all fwmark 0xca lookup 202
32221:	from all fwmark 0xc9 lookup 201
32222:	from all fwmark 0xca lookup 202
32223:	from all fwmark 0xc9 lookup 201
32224:	from all fwmark 0xca lookup 202
32225:	from all fwmark 0xc9 lookup 201
32226:	from all fwmark 0xca lookup 202
[edit]

 vyos@router# ip rule | grep "fwmark 0xc" | wc -l
556
[edit]
vyos@router# 
[edit]
vyos@router# ip rule | grep "fwmark 0xc" | wc -l
558
[edit]
vyos@router# 

```

3. If an interfaces in the `down` (link down) state (suspend in GNS3), the service restarts with `ERROR: Failed to apply WLB nftables config` continuously
```
set load-balancing wan interface-health eth0 nexthop '10.0.1.1'
set load-balancing wan interface-health eth1 nexthop '10.0.2.1'
set load-balancing wan rule 10 inbound-interface 'eth2'
set load-balancing wan rule 10 interface eth0
set load-balancing wan rule 10 interface eth1


vyos@router# run show int
Codes: S - State, L - Link, u - Up, D - Down, A - Admin Down
Interface    IP Address          MAC                VRF        MTU  S/L    Description
-----------  ------------------  -----------------  -------  -----  -----  -------------
eth0         10.0.1.2/24         0c:67:b0:98:00:00  default   1500  u/D    ISP-01
eth1         10.0.2.2/24         0c:67:b0:98:00:01  default   1500  u/u    ISP-02

```
logs:
```
Jan 17 11:16:15 router vyos-configd[654]: commit_scripts: ['load-balancing_wan']
Jan 17 11:16:15 router vyos-configd[654]: Received message: {"type": "node", "last": true, "data": "/usr/libexec/vyos/conf_mode/load-balancing_wan.py"}
Jan 17 11:16:15 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 11:16:15 router vyos-configd[654]: Sending reply: error_code 1 with output
Jan 17 11:16:15 router vyos-configd[654]: scripts_called: ['load-balancing_wan']
Jan 17 11:16:15 router systemd[1]: opt-vyatta-config-tmp-new_config_2810.mount: Deactivated successfully.
Jan 17 11:16:16 router commit[3046]: Successful change to active configuration by user vyos on /dev/pts/2

Jan 17 11:16:21 router python3[3001]: Notice: Waiting for commit to complete...
Jan 17 11:16:21 router python3[3001]: INFO: State change: eth0 -> False
Jan 17 11:16:21 router python3[3001]: ERROR: Failed to apply WLB nftables config
Jan 17 11:16:21 router systemd[1]: vyos-wan-load-balance.service: Deactivated successfully.
Jan 17 11:16:22 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 1.
Jan 17 11:16:22 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 11:16:22 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.

Jan 17 11:16:27 router python3[3090]: INFO: State change: eth0 -> False
Jan 17 11:16:27 router python3[3090]: ERROR: Failed to apply WLB nftables config
Jan 17 11:16:27 router systemd[1]: vyos-wan-load-balance.service: Deactivated successfully.
Jan 17 11:16:27 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 2.
Jan 17 11:16:27 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 11:16:27 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.

Jan 17 11:16:32 router python3[3112]: INFO: State change: eth0 -> False
Jan 17 11:16:32 router python3[3112]: ERROR: Failed to apply WLB nftables config
Jan 17 11:16:32 router systemd[1]: vyos-wan-load-balance.service: Deactivated successfully.
Jan 17 11:16:33 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 3.
Jan 17 11:16:33 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 11:16:33 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.

Jan 17 11:16:38 router python3[3134]: INFO: State change: eth0 -> False
Jan 17 11:16:38 router python3[3134]: ERROR: Failed to apply WLB nftables config
Jan 17 11:16:38 router systemd[1]: vyos-wan-load-balance.service: Deactivated successfully.
Jan 17 11:16:38 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 4.
Jan 17 11:16:38 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 11:16:38 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.

Jan 17 11:16:43 router python3[3156]: INFO: State change: eth0 -> False
Jan 17 11:16:43 router python3[3156]: ERROR: Failed to apply WLB nftables config
Jan 17 11:16:43 router systemd[1]: vyos-wan-load-balance.service: Deactivated successfully.
Jan 17 11:16:44 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 5.
Jan 17 11:16:44 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 11:16:44 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.

Jan 17 11:16:49 router python3[3178]: INFO: State change: eth0 -> False
Jan 17 11:16:49 router python3[3178]: ERROR: Failed to apply WLB nftables config
Jan 17 11:16:49 router systemd[1]: vyos-wan-load-balance.service: Deactivated successfully.
Jan 17 11:16:49 router systemd[1]: vyos-wan-load-balance.service: Scheduled restart job, restart counter is at 6.
Jan 17 11:16:49 router systemd[1]: Stopped vyos-wan-load-balance.service - VyOS WAN Load Balancer.
Jan 17 11:16:49 router systemd[1]: Started vyos-wan-load-balance.service - VyOS WAN Load Balancer.
^C

```

sever-sever commented on src/helpers/vyos-load-balancer.py:
```diff
@@ -0,0 +1,296 @@
+        run('ip route flush cache')
+
+        if 'flush_connections' in lb:
+            run('conntrack -F')
```
Use please `conntrack --delete`
We had bugs with flush https://github.com/vyos/vyatta-wanloadbalance/blob/5f6d38762d25267f974f8035ed79a0369df710c4/src/lbdecision.cc#L297-L300

https://vyos.dev/T1311

JeffWDH commented:
FYI, WLB config migration doesn't seem to be working. I've created https://vyos.dev/T7196
